src/safeposix/syscalls/net_calls.rs

```diff
@@ -1,5 +1,1505 @@
+                                }
+                            }
+                            mutmetadata.listening_port_set.insert(porttuple);
+
+                            sockfdobj.state = ConnState::LISTEN;
+
+                            //get or create the socket and bind it before listening
+                            let sid = Self::getsockobjid(sockfdobj);
+                            let locksock = mutmetadata.socket_object_table.get(&sid).unwrap().clone();
+                            let sockobj = locksock.read().unwrap();
+                            if let None = sockfdobj.localaddr {
+                                let bindret = sockobj.bind(&ladr);
+                                if bindret < 0 {
+                                    match Errno::from_discriminant(interface::get_errno()) {
+                                        Ok(i) => {return syscall_error(i, "listen", "The libc call to bind within listen failed");},
+                                        Err(()) => panic!("Unknown errno value from socket bind within listen returned!"),
+                                    };
+                                }
+                            }
+                            let listenret = sockobj.listen(5); //default backlog in repy for whatever reason, we replicate it
+                            if listenret < 0 {
+                                let lr = match Errno::from_discriminant(interface::get_errno()) {
+                                    Ok(i) => syscall_error(i, "listen", "The libc call to listen failed!"),
+                                    Err(()) => panic!("Unknown errno value from socket listen returned!"),
+                                };
+                                mutmetadata.listening_port_set.remove(&mux_port(ladr.addr().clone(), ladr.port(), sockfdobj.domain, TCPPORT));
+                                sockfdobj.state = ConnState::CONNECTED;
+                                return lr;
+                            };
+                            return 0;
+                        }
+                    }
+                }
+
+                _ => {
+                    return syscall_error(Errno::ENOTSOCK, "listen", "file descriptor refers to something other than a socket");
+                }
+            }
+        } else {
+            return syscall_error(Errno::EBADF, "listen", "invalid file descriptor");
+        }
+    }
+
+    pub fn netshutdown_syscall(&self, fd: i32, how: i32) -> i32 {
+        let mut fdtable = self.filedescriptortable.write().unwrap();
+        match how {
+            SHUT_RD => {
+                return syscall_error(Errno::EOPNOTSUPP, "netshutdown", "partial shutdown read is not implemented");
+            }
+            SHUT_WR => {
+                return Self::_cleanup_socket(self, fd, true, &mut fdtable);
+            }
+            SHUT_RDWR => {
+                //BUG:: need to check for duplicate entries
```

Comment:
Just leaving this for now?

Reply:
This has been fixed for months with my socket object reference counting.